mesonbuild/options.py

```diff
@@ -250,19 +250,23 @@ def evolve(self,
                          subproject if subproject != _BAD_VALUE else self.subproject, # None is a valid value so it can'the default value in method declaration.
                          machine if machine is not None else self.machine)
 
-    def as_root(self) -> 'OptionKey':
-        """Convenience method for key.evolve(subproject='')."""
-        if self.subproject is None or self.subproject == '':
+    def as_root(self) -> OptionKey:
+        """Convenience method for key.evolve(subproject='') if subproject is not global."""
+        if not self.subproject:
             return self
-        return self.evolve(subproject='')
+        return OptionKey(self.name, subproject='', machine=self.machine)
+
+    def as_toplevel(self) -> OptionKey:
```

Comment:
I don't think `as_toplevel` is clear here, I immediately thought "oh as_root means subproject='' and as_toplevel means subproject=None".

I'm also not sure why `as_root` doesn't universally set subproject='', since that's what it was added for, and I'm not convined the change to make it special case `subproject is None` is valid, that sounds like the caller isn't using the function correctly and just broke the function instead of actually checking that they were doing what they meant.